File: NeteaseM2DClient/app/src/main/java/com/aoihosizora/neteasem2dclient/MainActivity.java
package com.aoihosizora.neteasem2dclient;

import android.Manifest;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.media.MediaMetadata;
import android.media.session.PlaybackState;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.Toast;

import com.jwsd.libzxing.OnQRCodeScanCallback;
import com.jwsd.libzxing.QRCodeManager;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;
import java.util.Map;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class MainActivity extends AppCompatActivity {

    public static final int REQUEST_NETWORK_PERMISSION_CODE = 1;
    public static final int REQUEST_CAMERA_PERMISSION_CODE = 2;

    public static final String QRCODE_MAGIC = "NETEASE_M2D://";

    private SharedPreferences ipSP, portSP;

    private Intent ServiceIntent;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);
        ServiceIntent = new Intent(this, MainService.class);

        // 服务事件
        MainService.mainEvent = new MainEvent() {

            @Override
            @UiThread
            public void onDisconnect() {
                showAlert(getString(R.string.alert_break_connect), null);
                on_ui_stop();
                on_controller_stop();
            }

            @Override
            @UiThread
            public void onNoSession() {
                showAlert(getString(R.string.alert_no_session), null);
                on_ui_stop();
                on_controller_stop();
            }

            @Override
            @UiThread
            public void onSessionDestroy() {
                showAlert(getString(R.string.alert_session_destroy), null);
                on_ui_stop();
                on_controller_stop();
            }
        };

        // 访问权限
        if (!isNotificationListenersEnabled()) {
            showAlert(getString(R.string.alert_check_per), (d, i) -> {
                Intent intent = new Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS");
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
            });
        }

        ipSP = getSharedPreferences("ip", Context.MODE_PRIVATE);
        portSP = getSharedPreferences("port", Context.MODE_PRIVATE);

        // 自动完成输入框
        initAutoCompleteEdit();
    }

    /**
     * 通知へのアクセス
     */
    public boolean isNotificationListenersEnabled() {
        String pkgName = getPackageName();
        final String flat = Settings.Secure.getString(getContentResolver(), "enabled_notification_listeners");
        if (!TextUtils.isEmpty(flat)) {
            final String[] names = flat.split(":");
            for (String name : names) {
                final ComponentName cn = ComponentName.unflattenFromString(name);
                if (cn != null && TextUtils.equals(pkgName, cn.getPackageName()))
                    return true;
            }
        }
        return false;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        switch (requestCode) {
            case REQUEST_NETWORK_PERMISSION_CODE:
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED)
                    Toast.makeText(this, R.string.tst_auth_failed, Toast.LENGTH_LONG).show();
                break;
            case REQUEST_CAMERA_PERMISSION_CODE:
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED)
                    Toast.makeText(this, R.string.tst_auth_failed, Toast.LENGTH_LONG).show();
                else
                    on_btn_qrCode_clicked();
                break;
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    @BindView(R.id.id_btn_service)
    Button m_btn_service;
    @BindView(R.id.id_btn_qrCode)
    Button m_btn_qrCode;
    @BindView(R.id.id_btn_update)
    Button m_btn_update;

    @BindView(R.id.id_edt_ip)
    AutoCompleteTextView m_edt_ip;
    @BindView(R.id.id_edt_port)
    AutoCompleteTextView m_edt_port;

    /**
     * 自动补全文本框 弹出
     */
    @OnClick({R.id.id_edt_ip, R.id.id_edt_port})
    void on_edt_auto_completed_clicked(View view) {
        ((AutoCompleteTextView) view).showDropDown();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        QRCodeManager.getInstance().with(this).onActivityResult(requestCode, resultCode, data);
    }

    /**
     * 扫描二维码
     */
    @OnClick(R.id.id_btn_qrCode)
    void on_btn_qrCode_clicked() {

        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{
                Manifest.permission.CAMERA,
                Manifest.permission.WRITE_EXTERNAL_STORAGE
            }, REQUEST_CAMERA_PERMISSION_CODE);
            return;
        }

        QRCodeManager.getInstance()
            .with(this)
            .scanningQRCode(new OnQRCodeScanCallback() {
                @Override
                public void onCompleted(String result) {
                    try {
                        // 无 MAGIC
                        if (!result.startsWith(QRCODE_MAGIC))
                            throw new Exception();
                        result = result.substring(QRCODE_MAGIC.length());
                        // 无端口
                        if (!result.contains(":"))
                            throw new Exception();

                        // 分割
                        String[] sp = result.split(":");
                        m_edt_ip.setText(sp[0]);
                        m_edt_port.setText(sp[1]);
                    } catch (Exception ex) {
                        ex.printStackTrace();
                        Toast.makeText(MainActivity.this, getString(R.string.tst_qr_error), Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onError(Throwable errorMsg) {
                    Toast.makeText(MainActivity.this, getString(R.string.tst_qr_failed), Toast.LENGTH_SHORT).show();
                }

                @Override
                public void onCancel() {
                }
            });
    }

    /**
     * 开始结束服务
     */
    @OnClick(R.id.id_btn_service)
    @UiThread
    void on_btn_service_clicked() {
        if (m_btn_service.getText().equals(getString(R.string.btn_start))) {
            // 开启服务
            final String ip_str = m_edt_ip.getText().toString();
            final String port_str = m_edt_port.getText().toString();

            String ip_re = "^((25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$";
            String port_re = "([0-9]|[1-9]\\d{1,3}|[1-5]\\d{4}|6[0-4]\\d{4}|65[0-4]\\d{2}|655[0-2]\\d|6553[0-5])";

            if (!ip_str.matches(ip_re))
                showAlert(getString(R.string.alert_ip_format), null);
            else if (!port_str.matches(port_re))
                showAlert(getString(R.string.alert_port_format), null);
            else
                connect();
        } else {
            // 断开服务
            disconnect();
            new Thread(() -> {
                if (!SendService.sendMsg("{\"isDestroyed\": \"true\"}")) {
                    if (MainService.mainEvent != null)
                        MainService.mainEvent.onDisconnect();
                }
            }).start();
        }
    }

    /**
     * 手动更新
     */
    @OnClick(R.id.id_btn_update)
    void on_btn_update_clicked() {
        if (MainService.mediaController != null && MainService.mediaCallBack != null) {
            PlaybackState state = MainService.mediaController.getPlaybackState();
            MediaMetadata metadata = MainService.mediaController.getMetadata();

            if (state != null && metadata != null && state.getState() == PlaybackState.STATE_PLAYING) {
                MainService.mediaCallBack.onMetadataChanged(MainService.mediaController.getMetadata());
                MainService.mediaCallBack.onPlaybackStateChanged(MainService.mediaController.getPlaybackState());
            }
        }
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * 格式正确，连接
     */
    @UiThread
    private void connect() {
        // 判断网路权限
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.INTERNET) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{
                Manifest.permission.INTERNET
            }, REQUEST_NETWORK_PERMISSION_CODE);
        }

        final boolean[] isCanceled = new boolean[]{false};
        final ProgressDialog progressDialog = new ProgressDialog(this);
        progressDialog.setMessage(getString(R.string.progress_linking));
        progressDialog.setCancelable(true);
        progressDialog.setOnCancelListener((d) -> isCanceled[0] = true);
        progressDialog.show();

        new Thread(() -> {
            SendService.ip = m_edt_ip.getText().toString();
            SendService.port = Integer.parseInt(m_edt_port.getText().toString());

            if (!SendService.ping()) {

                if (isCanceled[0]) return;

                // 连接不通
                runOnUiThread(() -> {
                    progressDialog.dismiss();
                    showAlert(getString(R.string.alert_connect_failed), null);
                    on_ui_stop();
                });
                return;
            }

            if (isCanceled[0]) return;

            runOnUiThread(() -> {
                progressDialog.dismiss();
                Toast.makeText(MainActivity.this, R.string.tst_connect_success, Toast.LENGTH_SHORT).show();

                // 保存
                saveIP(SendService.ip);
                savePort(SendService.port);
                initAutoCompleteEdit();

                // 开始服务
                on_controller_stop();
                startService(ServiceIntent);
                on_ui_starting();
            });
        }).start();
    }

    /**
     * 断开连接
     */
    @UiThread
    private void disconnect() {
        on_ui_stop();
        on_controller_stop();
    }

    /**
     * 连接界面更新
     */
    @UiThread
    private void on_ui_starting() {
        m_btn_service.setText(getString(R.string.btn_stop));
        m_edt_ip.setEnabled(false);
        m_edt_port.setEnabled(false);
        m_btn_qrCode.setEnabled(false);
        m_btn_update.setEnabled(true);
    }

    /**
     * 断开连接界面更新
     */
    @UiThread
    private void on_ui_stop() {
        m_btn_service.setText(getString(R.string.btn_start));
        m_edt_ip.setEnabled(true);
        m_edt_port.setEnabled(true);
        m_btn_qrCode.setEnabled(true);
        m_btn_update.setEnabled(false);
    }

    /**
     * 主动注销通知，(不能写在 onDestroy)
     */
    private void on_controller_stop() {
        if (MainService.mediaController != null)
            MainService.mediaController.unregisterCallback(MainService.mediaCallBack);

        MainService.mediaController = null;
        MainService.mediaCallBack = null;

        stopService(ServiceIntent);
    }

    @Override
    protected void onDestroy() {
        on_controller_stop();
        super.onDestroy();
    }

    private void showAlert(String msg, DialogInterface.OnClickListener listener) {
        new AlertDialog.Builder(this)
            .setTitle(R.string.alert_title)
            .setMessage(msg)
            .setPositiveButton(R.string.alert_pos, listener)
            .show();
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    /**
     * 获得当前时间
     *
     * @return yyyy-MM-dd-HH-mm-ss
     */
    private String getCurrentTime() {
        SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd-HH-mm-ss", Locale.CHINA);
        Date date = new Date(System.currentTimeMillis());
        return simpleDateFormat.format(date);
    }

    /**
     * 保存 IP 至 SharedPreferences
     */
    private void saveIP(String ip) {
        boolean isSaved = false;
        for (Map.Entry<String, ?> entry : ipSP.getAll().entrySet()) {
            if (entry.getValue().equals(ip)) {
                isSaved = true;
                break;
            }
        }
        if (!isSaved) {
            SharedPreferences.Editor ipSPEdit = ipSP.edit();
            ipSPEdit.putString(getCurrentTime(), ip);
            ipSPEdit.apply();
        }
    }

    /**
     * 保存 Port 至 SharedPreferences
     */
    private void savePort(int port) {
        boolean isSaved = false;
        for (Map.Entry<String, ?> entry : portSP.getAll().entrySet()) {
            try {
                if (Integer.parseInt(entry.getValue().toString()) == port) {
                    isSaved = true;
                    break;
                }
            } catch (Exception ex) {
                ex.printStackTrace();
            }
        }
        if (!isSaved) {
            SharedPreferences.Editor ipSPEdit = portSP.edit();
            ipSPEdit.putInt(getCurrentTime(), port);
            ipSPEdit.apply();
        }
    }

    /**
     * 获取存储的 IP
     */
    private String[] getSavedIP() {
        ArrayList<String> ret = new ArrayList<>();
        for (Map.Entry<String, ?> entry : ipSP.getAll().entrySet()) {
            ret.add(entry.getValue().toString());
        }
        return ret.toArray(new String[0]);
    }

    /**
     * 获取存储的 Port
     */
    private String[] getSavedPort() {
        ArrayList<String> ret = new ArrayList<>();
        for (Map.Entry<String, ?> entry : portSP.getAll().entrySet()) {
            ret.add(entry.getValue().toString());
        }
        return ret.toArray(new String[0]);
    }

    /**
     * 初始化自动完成框
     */
    private void initAutoCompleteEdit() {
        ArrayAdapter ipAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, getSavedIP());
        ArrayAdapter portAdapter = new ArrayAdapter<>(this, android.R.layout.simple_list_item_1, getSavedPort());
        m_edt_ip.setAdapter(ipAdapter);
        m_edt_port.setAdapter(portAdapter);
        ipAdapter.notifyDataSetChanged();
        portAdapter.notifyDataSetChanged();
    }
}

File: android-client/app/src/main/java/com/aoihosizora/ncmlatwclient/MainActivity.kt
package com.aoihosizora.ncmlatwclient

import android.Manifest
import android.app.*
import android.content.ComponentName
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.provider.Settings
import android.text.TextUtils
import android.view.WindowManager
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Toast
import androidx.annotation.UiThread
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.app.NotificationCompat
import com.jwsd.libzxing.OnQRCodeScanCallback
import com.jwsd.libzxing.QRCodeManager
import kotlinx.android.synthetic.main.activity_main.*
import java.text.SimpleDateFormat
import java.util.*

class MainActivity : AppCompatActivity() {

    companion object {
        private const val QRCODE_MAGIC = "NCMLATW-"
        private const val REQUEST_CAMERA_PERMISSION_CODE = 1
        private var SHOWING_NO_SESSION_DIALOG = false
        private var SHOWING_DISCONNECT_DIALOG = false
    }

    private lateinit var adtIP: ArrayAdapter<String>
    private lateinit var adtPort: ArrayAdapter<String>
    private var lastBackPressedTime: Date? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initUI()
        initPermission()
        initListener()
    }

    private fun initUI() {
        adtIP = ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, getDataFromSP("ip"))
        adtPort = ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, getDataFromSP("port"))
        edt_ip.setAdapter(adtIP)
        edt_port.setAdapter(adtPort)
        adtIP.notifyDataSetChanged()
        adtPort.notifyDataSetChanged()
    }

    private fun initPermission() {
        var granted = false
        val flat = Settings.Secure.getString(contentResolver, "enabled_notification_listeners")
        if (!TextUtils.isEmpty(flat)) {
            for (name in flat.split(":")) {
                val cn = ComponentName.unflattenFromString(name)
                if (cn != null && TextUtils.equals(packageName, cn.packageName)) {
                    granted = true
                    break
                }
            }
        }

        if (!granted) {
            showAlertDialog("授权", "该应用需要授权通知访问，请授权。", false) { _, _ ->
                // 通知へのアクセス
                val intent = Intent("android.settings.ACTION_NOTIFICATION_LISTENER_SETTINGS")
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                startActivity(intent)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
    }

    override fun onPause() {
        window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        super.onPause()
    }

    override fun onDestroy() {
        processStopService()
        super.onDestroy()
    }

    override fun onBackPressed() {
        val current = Calendar.getInstance().time
        if (lastBackPressedTime == null || current.time - lastBackPressedTime!!.time > 2000) {
            Toast.makeText(this, "再按一次退出应用", Toast.LENGTH_SHORT).show()
            lastBackPressedTime = current
        } else {
            super.onBackPressed()
        }
    }

    private fun initListener() {
        MainService.eventListener = EventListener()

        edt_ip.setOnClickListener { v -> (v as AutoCompleteTextView).showDropDown() }
        edt_port.setOnClickListener { v -> (v as AutoCompleteTextView).showDropDown() }

        btn_qrcode.setOnClickListener { scanQrcode() }
        btn_manual.setOnClickListener { updateManually() }
        btn_clearLog.setOnClickListener { clearLog() }
        btn_service.setOnClickListener { if (btn_service.text == "启动服务") startService() else stopService() }
    }

    // ===============
    // service related
    // ===============

    @UiThread
    private fun startService() {
        // check address
        val ip = edt_ip.text.toString()
        val portStr = edt_port.text.toString()
        if (!SocketUtils.checkIPString(ip)) {
            showAlertDialog("启动服务", "IP 地址格式有误。")
            return
        }
        if (!SocketUtils.checkPortString(portStr)) {
            showAlertDialog("启动服务", "无效的端口。")
            return
        }
        val port = Integer.parseInt(portStr)

        // show progress dlg
        var isCanceled = false
        val dlg = ProgressDialog(this)
        dlg.setMessage("正在连接...")
        dlg.setCancelable(true)
        dlg.setOnCancelListener { isCanceled = true }
        dlg.show()

        Thread {
            // start connect
            SocketUtils.storeAddress(ip, port)
            val ok = SocketUtils.ping()
            if (isCanceled) {
                return@Thread
            }

            // start service
            runOnUiThread {
                dlg.dismiss()
                if (!ok) {
                    showAlertDialog("启动服务", "无法连接到给定地址。")
                } else {
                    processStartService(ip, portStr)
                }
            }
        }.start()
    }

    @UiThread
    private fun stopService() {
        // stop service
        processStopService()

        // send destroyed
        val dto = DestroyedDto(true)
        dto.toJSON()?.let { sendToServer(it.toString(), failAlert = false) }
    }

    /**
     * Process start service, includes: start service, update ui, save data to sp.
     */
    @UiThread
    private fun processStartService(ip: String, port: String) {
        // start service
        if (MainService.isRunning(this)) {
            appendLog("Before start service, stopping service...")
            MainService.mediaCallback?.let { MainService.mediaController?.unregisterCallback(it) }
            MainService.mediaController = null
            MainService.mediaCallback = null
            stopService(Intent(this, MainService::class.java))
        }
        if (!MainService.isRunning(this)) {
            appendLog("Starting service...")
            startService(Intent(this, MainService::class.java))
        }

        // save sp and update adapter
        if (saveDataToSP("ip", ip)) {
            adtIP.add(ip)
            adtIP.notifyDataSetChanged()
        }
        if (saveDataToSP("port", port)) {
            adtPort.add(port)
            adtPort.notifyDataSetChanged()
        }

        // update ui
        btn_service.text = "停止服务"
        edt_ip.isEnabled = false
        edt_port.isEnabled = false
        btn_qrcode.isEnabled = false
        btn_manual.isEnabled = true

        // setup notification
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channel = NotificationChannel("ncmlatw", "服务状态", NotificationManager.IMPORTANCE_DEFAULT)
        nm.createNotificationChannel(channel)
        val contentIntent = PendingIntent.getActivity(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT)
        val actionIntent = PendingIntent.getService(this, 0, Intent(this, NotificationIntentService::class.java), PendingIntent.FLAG_UPDATE_CURRENT)
        val notification = NotificationCompat.Builder(this, "ncmlatw")
            .setAutoCancel(false)
            .setOngoing(true)
            .setContentTitle("服务已启动")
            .addAction(R.mipmap.ic_launcher, "手动更新", actionIntent)
            .setContentText("正在与 PC 端保持着连接")
            .setWhen(System.currentTimeMillis())
            .setSmallIcon(R.mipmap.ic_launcher)
            .setContentIntent(contentIntent)
            .build()
        nm.notify(1, notification)
    }

    /**
     * Process stop service, includes: stop service, update ui.
     */
    @UiThread
    private fun processStopService() {
        // stop service
        if (MainService.isRunning(this)) {
            appendLog("Stopping service...")
            MainService.mediaCallback?.let { MainService.mediaController?.unregisterCallback(it) }
            MainService.mediaController = null
            MainService.mediaCallback = null
            stopService(Intent(this, MainService::class.java))
        }

        // update ui
        btn_service.text = "启动服务"
        edt_ip.isEnabled = true
        edt_port.isEnabled = true
        btn_qrcode.isEnabled = true
        btn_manual.isEnabled = false

        // remove notification
        val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        nm.cancel(1)
    }

    // =====================
    // event respond related
    // =====================

    inner class EventListener : MainService.EventListener {

        @UiThread
        override fun onNoSession() {
            showAlertDialog("启动服务", "没有打开网易云音乐 app。")
            processStopService()
        }

        @UiThread
        override fun onSessionDestroyed() {
            if (!SHOWING_NO_SESSION_DIALOG) {
                SHOWING_NO_SESSION_DIALOG = true
                showAlertDialog("服务", "网易云音乐 app 已经关闭。", false) { _, _ -> SHOWING_NO_SESSION_DIALOG = false }
            }
            processStopService()

            val dto = DestroyedDto(true)
            dto.toJSON()?.let { sendToServer(it.toString()) }
        }

        @UiThread
        override fun onPlaybackStateChanged(isPlaying: Boolean, position: Double) {
            val dto = PlaybackStateDto(isPlaying, position)
            dto.toJSON()?.let { sendToServer(it.toString()) }
        }

        @UiThread
        override fun onMetadataChanged(title: String, artist: String, album: String, duration: Double) {
            val dto = MetadataDto(title, artist, album, duration)
            dto.toJSON()?.let { sendToServer(it.toString()) }
        }

        @UiThread
        override fun onServiceLifetime(text: String) {
            appendLog("Service: $text")
        }
    }

    /**
     * Append text to log, send text to server with retrying, show alert dialog when failed and switch on.
     */
    @UiThread
    private fun sendToServer(text: String, failAlert: Boolean = true) {
        appendLog(text)

        Thread {
            for (i in 0..2) { // retry twice
                if (SocketUtils.send(text)) {
                    return@Thread
                }
            }
            if (failAlert) {
                runOnUiThread {
                    if (!SHOWING_DISCONNECT_DIALOG) {
                        SHOWING_DISCONNECT_DIALOG = true
                        showAlertDialog("服务", "与 PC 端的连接已断开。", false) { _, _ -> SHOWING_DISCONNECT_DIALOG = false }
                    }
                    processStopService()
                }
            }
        }.start()
    }

    // ================
    // other actions...
    // ================

    @UiThread
    private fun updateManually() {
        MainService.mediaController?.let { it1 ->
            MainService.mediaCallback?.let { it2 ->
                val state = it1.playbackState
                val metadata = it1.metadata
                if (state != null && metadata != null) {
                    it2.onMetadataChanged(metadata)
                    it2.onPlaybackStateChanged(state)
                }
            }
        }
    }

    class NotificationIntentService : IntentService("NcmlAtwIntentService") {
        override fun onHandleIntent(intent: Intent?) {
            MainService.mediaController?.let { it1 ->
                MainService.mediaCallback?.let { it2 ->
                    val state = it1.playbackState
                    val metadata = it1.metadata
                    if (state != null && metadata != null) {
                        it2.onMetadataChanged(metadata)
                        it2.onPlaybackStateChanged(state)
                    }
                }
            }
        }
    }

    @UiThread
    private fun scanQrcode() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA), REQUEST_CAMERA_PERMISSION_CODE)
            return
        }

        QRCodeManager.getInstance()
            .with(this)
            .scanningQRCode(object : OnQRCodeScanCallback {
                override fun onCompleted(result: String?) {
                    if (result == null) {
                        return
                    }
                    if (!result.startsWith(QRCODE_MAGIC)) {
                        showAlertDialog("扫描二维码", "不支持的二维码。")
                        return
                    }
                    val addrSp = result.substring(QRCODE_MAGIC.length).split(":")
                    if (addrSp.size != 2) {
                        showAlertDialog("扫描二维码", "不支持的二维码。")
                        return
                    }

                    val ip = addrSp[0]
                    val portStr = addrSp[1]
                    if (!SocketUtils.checkIPString(ip) || !SocketUtils.checkPortString(portStr)) {
                        showAlertDialog("扫描二维码", "二维码包含格式错误的 IP 地址和端口。")
                        return
                    }

                    edt_ip.setText(ip)
                    edt_port.setText(portStr)
                }

                override fun onError(error: Throwable?) {
                    showAlertDialog("扫描二维码", "扫描二维码失败。")
                }

                override fun onCancel() {}
            })
    }

    // =========
    // others...
    // =========

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        QRCodeManager.getInstance().with(this).onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when (requestCode) {
            REQUEST_CAMERA_PERMISSION_CODE -> {
                if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                    showAlertDialog("授权", "相机授权失败。")
                } else {
                    scanQrcode()
                }
            }
        }
    }

    @UiThread
    private fun showAlertDialog(title: String, message: String, cancelable: Boolean = true, okCallback: ((DialogInterface, Int) -> Unit)? = null) {
        AlertDialog.Builder(this)
            .setTitle(title)
            .setMessage(message)
            .setCancelable(cancelable)
            .setPositiveButton("确定", okCallback)
            .show()
    }

    @UiThread
    private fun clearLog() {
        tv_log.text = ""
    }

    @UiThread
    private fun appendLog(text: String) {
        val fmt = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.CHINA)
        val now = fmt.format(Calendar.getInstance().time)
        tv_log.append("\n[$now] $text")
        val scrollAmount = tv_log.layout.getLineTop(tv_log.lineCount) - tv_log.height
        if (scrollAmount > 0) {
            tv_log.scrollTo(0, scrollAmount)
        } else {
            tv_log.scrollTo(0, 0)
        }
    }

    private fun getDataFromSP(name: String): Array<String> {
        val sp = getSharedPreferences(name, Context.MODE_PRIVATE)
        val out = arrayListOf<String>()
        for (entry in sp.all.entries) {
            out.add(entry.value.toString())
        }
        return out.toArray(arrayOf<String>())
    }

    private fun saveDataToSP(name: String, value: String): Boolean {
        val sp = getSharedPreferences(name, Context.MODE_PRIVATE)
        if (!getDataFromSP(name).contains(value)) {
            val fmt = SimpleDateFormat("yyyyMMddHHmmss", Locale.CHINA)
            val token = fmt.format(Calendar.getInstance().time)
            val edt = sp.edit()
            edt.putString(token, value)
            edt.apply()
            return true
        }
        return false
    }
}
